tests/nn/data_parallel/test_fsdp_memory.py

```diff
@@ -0,0 +1,243 @@
+    """Collect memory allocated values in a result dict in MB"""
+    result[prefix] = round(torch.cuda.memory_allocated() / 1024 / 1024)
+
+
+class Model(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.stem = nn.Sequential(nn.Conv2d(3, 64, kernel_size=3), nn.BatchNorm2d(64), nn.ReLU(inplace=True))
+        self.blocks = nn.Sequential(
+            nn.Conv2d(64, 128, kernel_size=5, padding=2),
+            nn.BatchNorm2d(128),
+            nn.ReLU(inplace=True),
+            nn.Conv2d(128, 128, kernel_size=5, padding=2),
+            nn.BatchNorm2d(128),
+            nn.ReLU(inplace=True),
+            nn.Conv2d(128, 128, kernel_size=5, padding=2),
+            nn.BatchNorm2d(128),
+            nn.ReLU(inplace=True),
+            nn.AdaptiveAvgPool2d(output_size=(1, 1)),
+            nn.Flatten(),
+        )
+        self.head = nn.Linear(128, 10)
+
+    def forward(self, x):
+        return self.head(self.blocks(self.stem(x)))
+
+
+def create_model(with_fsdp, with_checkpoint):
+    model = Model()
+    if with_fsdp:
+        model.stem = auto_wrap_bn(model.stem, single_rank_pg=False)
+        model.blocks = auto_wrap_bn(model.blocks, single_rank_pg=False)
+        if with_checkpoint:
+            model.blocks = checkpoint_wrapper(model.blocks)
+        model.stem = to_fsdp(model.stem)
+        model.blocks = to_fsdp(model.blocks)
+        model.head = to_fsdp(model.head)
+    else:
+        if with_checkpoint:
+            model.blocks = checkpoint_wrapper(model.blocks)
+    return model
+
+
+def _distributed_worker(gpu_id, world_size, with_fsdp, with_checkpoint, filename, filename_rpc, expected):
+    torch.cuda.set_device(gpu_id)
+
+    rank = gpu_id
+    result = dist_init(rank, world_size, filename, filename_rpc)
+    assert result, "Dist init failed"
+
+    torch.manual_seed(0)
+    torch.backends.cudnn.deterministic = True
+    batch = torch.randn(size=(2, 3, 224, 224)).cuda()
+
+    model = create_model(with_fsdp, with_checkpoint)
+    model = model.cuda()
+    if with_fsdp:
+        model = to_fsdp(model)
+    else:
+        model = DistributedDataParallel(model, device_ids=[gpu_id], bucket_cap_mb=500)
+
+    criterion = nn.MSELoss()
+    optimizer = optim.SGD(model.parameters(), lr=1e-4)
+
+    results = {}
+    for iteration in range(3):
+        get_cur_mem(gpu_id, results, f"iter {iteration}: start")
+
+        out = model(batch)
+        get_cur_mem(gpu_id, results, f"iter {iteration}: after fwd")
+
+        out = sum(o.sum() for o in out[0])
+        fake_loss = criterion(out, torch.tensor(0.0).cuda())
+        get_cur_mem(gpu_id, results, f"iter {iteration}: after loss")
+
+        fake_loss.backward()
+        get_cur_mem(gpu_id, results, f"iter {iteration}: after bwd")
+
+        optimizer.step()
+        get_cur_mem(gpu_id, results, f"iter {iteration}: after step")
+
+        # It is important to use the loop below, not optimizer.zero_grad() to reclaim memory.
```

Comment:
This seems to be very important for FSDP. Without it FSDP end up with >60MB after the first iter and DDP ends up with around 12MB. I think it is good for everyone to do it this way anyway, so I don't think we need to fix this in FSDP.

Reply:
just a minor comment: we usually do `optimizer.zero_grad() -> optimizer.step()` in this order not the other way around.

Reply:
see below. we can use model.zero_grad(set_to_none=True) as well. I recommended `optimizer.step() -> model.zero_grad()` since this will free the memory sooner than the other way around.